src/poetry/console/commands/add.py

```diff
@@ -54,7 +54,12 @@ class AddCommand(InstallerCommand, InitCommand):
             flag=False,
             multiple=True,
         ),
-        option("optional", None, "Add as an optional dependency."),
+        option(
+            "optional",
+            None,
+            "Add as an optional dependency to an extra.",
+            flag=False,
+        ),
```

Comment:
If I'm reading this change correctly, the syntax for adding dependencies in Poetry would change when this is committed, correct?

```bash
$ # Before, to add pytest as an optional dependency
$ poetry add --optional pytest
$ # Now, to add pytest to the "dev" extra as a dependency
$ poetry add --optional dev pytest
```

...Do I have that right? If so, is there a chance that users might be thrown by the syntax _changing_ like that from one release to the next? Having to check which Poetry version you're working with to get the syntax right is kind of a bummer. To say nothing of anyone who's scripted or created aliases for Poetry commands, and has their scripts break on update because the syntax is wrong.

Would it be possible to, instead of changing `--optional` like that, instead provide a _new_ syntax, and deprecate the old `--optional`? (It doesn't entirely make sense anymore, given that its extras support is kind of shoe-horned in.)

It's too bad `poetry add` already has an `--extras` flag, since the natural syntax for the new command would seem to be,
```bash
$ poetry add --extra dev pytest
```

...But given the existence of `--extras` as well, that seems far too confusing.

Perhaps a new subcommand would even be better?

```bash
$ # perhaps...
$ poetry extra dev --add pytest
$ # or...
$ poetry extra add dev pytest
$ poetry extra add --to dev pytest
```

The latter would also leave room to add additional `extra` subcommands, like:
* `poetry extra list`
* `poetry extra info`
* `poetry extra install`, perhaps even

...All of that is outside the scope of this PR, of course. I'm just a bit concerned about the semantics of a long-standing command like `poetry add --optional` suddenly changing "overnight".

Reply:
Another possibility, perhaps `poetry add --optional` _without_ an extra specified could just add to a default extra, like `dev`? That way,

```bash
$ poetry add --optional pytest
```
could still work, and just become equivalent to `poetry add --optional dev pytest`. (It could even display a message warning that the old syntax is deprecated, and will be removed in the future.)

Reply:
> ...Do I have that right? If so, is there a chance that users might be thrown by the syntax changing like that from one release to the next?

fwiw I think this will be shipped in a major release, so a few breaking changes can be expected

Reply:
Yes, PEP 621 support will only come in a major release. We still try to introduce as few breaking changes as possible but we always have to weigh up whether avoiding a breaking change requires too much maintenance effort or results in confusing UI for new users - especially in a major release where breaking changes should be expected.

> Another possibility, perhaps `poetry add --optional` _without_ an extra specified could just add to a default extra, like `dev`?

I am not convinced that this is better than failing. It might be surprising that dependencies are added to a default extra. I think sometimes it is better to fail loudly than to silently do something unexpected.

> Perhaps a new subcommand would even be better?

> The latter would also leave room to add additional extra subcommands, like

That might make sense but it would delay the next release even further. We can still introduce such a subcommand later and deprecate `poetry add --optional` to improve UI. (Of course, that does not help to avoid a breaking change now.)

> All of that is outside the scope of this PR

Full ack.

Reply:
That's all fair and sensible, thanks @radoering (and @edgarrmondragon).